### PR TITLE
Add IsReadableStreamDisturbed predicate

### PR DESCRIPTION
**Pull request by tyoshino**
See the suggestion in the OP of #378.

### CODE REVIEW

**Comment by tyoshino** [comment]
The disturbed flag is never changed when the streams is closed. Is it assumed that this is implemented after removing the auto-release feature?

**Comment by yutakahirano** [comment]
FYI: https://github.com/yutakahirano/fetch-with-streams/pull/48#issuecomment-125172550

File: index.bs [hunk 1]
```diff
@@ -300,6 +300,11 @@ Instances of <code>ReadableStream</code> are created with the internal slots des
       state and queue of this stream
   </tr>
   <tr>
+    <td>\[[disturbed]]
+    <td>A boolean flag set to <b>true</b> when either <code>read()</code> or <code>cancel()</code> has been called on
```
Comment (domenic) [inline index.bs]: better phrased as "when the stream has been read from or canceled"

File: index.bs [hunk 2]
```diff
@@ -971,6 +978,16 @@ an assert).
   1. Return *true*.
 </emu-alg>
 
+<h4 id="is-readable-stream-disturbed" aoid="IsReadableStreamDisturbed">IsReadableStreamDisturbed ( stream )</h4>
+
+This abstract operation is meant to be called from other specifications that may wish to query whether or not a
```
Comment (domenic) [inline index.bs]: "...readable stream has ever been read from or canceled"

File: reference-implementation/run-tests.js [hunk 3]
```diff
@@ -2,12 +2,14 @@ const glob = require('glob');
...
 
 import ReadableStream from './lib/readable-stream';
+import { IsReadableStreamDisturbed } from './lib/readable-stream';
 import WritableStream from './lib/writable-stream';
 import ByteLengthQueuingStrategy from './lib/byte-length-queuing-strategy';
 import CountQueuingStrategy from './lib/count-queuing-strategy';
 import TransformStream from './lib/transform-stream';
 
 global.ReadableStream = ReadableStream;
+global.IsReadableStreamDisturbed = IsReadableStreamDisturbed;
```
Comment (domenic) [inline reference-implementation/run-tests.js]: I don't like this testing strategy because the global modifications are meant to be based on https://streams.spec.whatwg.org/#globals. I am not sure what the best strategy for testing it is though. I will try to work on something.

Comment (domenic) [inline reference-implementation/run-tests.js]: For example, such tests could never be run in browsers. Maybe they just should be left for other specs' test suites.

**Comment by domenic** [comment]
> The disturbed flag is never changed when the streams is closed. Is it assumed that this is implemented after removing the auto-release feature?

I don't think we need to remove auto-release. However, as @yutakahirano notes, I think it was important (for reasons I forget) to have the disturbed flag change even on closed streams. As such it should probably move up a few steps in the algorithms.

File: reference-implementation/run-tests.js (hunk 3 above)
Comment (tyoshino) [inline reference-implementation/run-tests.js]: So, should we remove this from run-tests.js and ... import in the test file?

File: index.bs (hunk 2 above)
Comment (tyoshino) [inline index.bs]: Done

File: index.bs (hunk 1 above)
Comment (tyoshino) [inline index.bs]: Done

**Comment by tyoshino** [comment]
Maybe one of the origins of the requirement is https://github.com/yutakahirano/fetch-with-streams/issues/37#issuecomment-108116997 ? Is it still active? https://github.com/whatwg/fetch/issues/61#issuecomment-123854971 Or even if it's active, it'll be covered by used flag?

**Comment by domenic** [comment]
Reviewing https://etherpad.mozilla.org/streams-f2f-july "Even more conservative strawman" I think the requirement comes from wanting .json() etc. to automatically make the stream disturbed, even for empty streams. The used flag is only for `new Request(otherRequest)`. This is important because it means that author code that acts similar to json() will also make the stream disturbed.

File: reference-implementation/run-tests.js (hunk 3 above)
Comment (domenic) [inline reference-implementation/run-tests.js]: Yes, that sounds good. And maybe move them to a separate abstract-ops.js or from-other-specs.js test file.

Comment (tyoshino) [inline reference-implementation/run-tests.js]: Done

**Comment by tyoshino** [comment]
Addressed the comment. Thanks.

> This is important because it means that author code that acts similar to json() will also make the stream disturbed.

OK. But an operation made on a released reader making the ex-associated stream disturbed sounds strange.

**Comment by domenic** [comment]
> But an operation made on a released reader making the ex-associated stream disturbed sounds strange.

Oh, you are right. It would mean that getting a reader for a closed/error stream would still have to hold on to the stream in order to maybe mark it disturbed.

This seems a bit silly. It means you cannot GC the stream if the reader is still alive, even though the stream is closed or errored.

But that is maybe not so bad since closed or errored streams do not have any data in their internal queue.

Also it would eliminate some internal complexity from the reader (the [[state]] and [[storedError]] fields).

And the silliness seems less bad than debating bodyUsed all over again.

What do you think?

**Comment by domenic** [comment]
Ping. What are your thoughts?

**Comment by yutakahirano** [comment]
I prefer dropping auto-release to introducing released-but-still-connected state.

**Comment by domenic** [comment]
That seems like a good compromise to me. @tyoshino do you want to do the PR or shall I?

**Comment by domenic** [comment]
Actually, I can do that; there is other work I want to do on top of it. I will send a PR later today.

**Comment by domenic** [comment]
OK, I have pushed to this PR with a base commit that removes automatic unlocking on close/error.

Two important questions:

- Should we now reject read() and closed when used on released readers, instead of treating as a closed/errored stream? That seems simpler.

- **Are we OK with breaking compat on this?** It would be a breaking change to Chrome's current behavior, and a change in an area of the spec we've called stable. I personally think it is unlikely to break things, but what do you think? /cc @yutakahirano @calvaris

**Comment by calvaris** [comment]
I think it is also unlikely that we break things.

Our current implementation is also based on closed/errored so it would mean to change our current implementation, which I'm not a big fan of but it wouldn't be a big change either, so I'm against but only barely and based on laziness :)

What I would do is securing the change of behavior if it happens with some tests ensuring that everything is as it should be and not as it was.

**Comment by domenic** [comment]
@calvaris thanks for weighing in! I made a bunch of test updates in https://github.com/whatwg/streams/commit/9945ae13eea9c1156d1feae3bfe665d9f7a88fc4 but if you can think of others we should add them. E.g. if you come up with test descriptions I'll write the tests :)

**Comment by yutakahirano** [comment]
I believe we cannot introduce IsDisturbed with satisfying all proposed requirements and all existing properties. I think dropping the auto-release is a tolerable way.

**Comment by domenic** [comment]
@yutakahirano well, we can introduce it without any backward-incompatible changes if we follow https://github.com/whatwg/streams/pull/385#issuecomment-128149814 and allow readers to retain references to their owner stream even after release.

But I take it you think that is bad?

**Comment by yutakahirano** [comment]
> But I take it you think that is bad?

Yes.

**Comment by domenic** [comment]
OK. Does anyone have any opinion on

> Should we now reject read() and closed when used on released readers, instead of treating as a closed/errored stream?

**Comment by domenic** [comment]
My opinion is that they should now reject. That will allow us to get rid of the internal slots [[state]] and [[storedError]]. The downside is that you would no longer get to treat released readers as closed/errored streams depending on their originating stream. Instead they would effectively behave as if they always came from errored streams. But I don't think that's a big deal; you shouldn't be messing with the reader after it's been released.

**Comment by tyoshino** [comment]
Sorry for delay. I'm fine with the plan. I don't come up with any reasonable use case to keep a reference to a reader while releasing the parent stream. Some code may be accidentally doing so, but could be easily fixed and as Domenic said it's not so harmful to hold an errored/closed stream (which is small) in addition to a reader.

I'm also fine with making read on released reader reject. I think currently people who are unaware of the auto release mechanism are not calling release(), and basically people who are aware of wouldn't read() on released one. Some may be using release() and also using released reader as a representation of closed stream that accepts read() and tell us that the stream is closed repeatedly. But ... maybe the number of users with such a use case is small, I guess.

**Comment by domenic** [comment]
OK. Updated with what I think should be final. Cannot build the spec until https://github.com/tabatkins/bikeshed/issues/466 is fixed, but ready for review.

File: index.bs [hunk 4]
```diff
@@ -855,7 +848,10 @@ Instances of <code>ReadableStreamReader</code> are created with the internal slo
   1. If IsReadableStreamReader(*this*) is *false*, throw a *TypeError* exception.
   1. If *this*@[[ownerReadableStream]] is *undefined*, return *undefined*.
   1. If *this*@[[readRequests]] is not empty, throw a *TypeError* exception.
-  1. Perform CloseReadableStreamReader(*this*).
+  1. If *this*@[[ownerReadableStream]]@[[state]] is `"readable"`, reject *this*@[[closedPromise]] with a *TypeError* exception.
+  1. Otherwise, let *this*@[[closedPromise]] be a new promise rejected with a *TypeError* exception.
```
Comment (tyoshino) [inline index.bs]: I thought "let" is used for declaring/substituting to local variables while set is used for substituting new value to a slot. I'm just misunderstanding your style rule or this "let" should be "set"?

Comment (tyoshino) [inline index.bs]: Do we really need to set a new rejected promise to [[closedPromise]] here?

File: index.bs [hunk 5]
```diff
@@ -973,8 +958,6 @@ an assert).
     1. Reject _readRequestPromise_ with _e_.
   1. Set _reader_@[[readRequests]] to a new empty List.
   1. Perform ReleaseReadableStreamReader(_reader_).
```
Comment (tyoshino) [inline index.bs]: Was the plan to keep auto-release for error?

Comment (tyoshino) [inline index.bs]: Oh, the impl is fixed. So, please just delete this line in the spec text :)

File: reference-implementation/lib/readable-stream.js [hunk 6]
```diff
@@ -562,13 +559,6 @@ function ReadFromReadableStreamReader(reader) {
   }
 }
 
-function ReleaseReadableStreamReader(reader) {
```
Comment (tyoshino) [inline reference-implementation/lib/readable-stream.js]: Let's remove ReleaseReadableStreamReader from the spec text too.

File: index.bs (hunk 4 above)
Comment (tyoshino) [inline index.bs]: Ah, I see. This new rejected promise and the TypeError at L851 are representing "already released". Then, looks good.

**Comment by tyoshino** [comment]
Review done!

**Comment by domenic** [comment]
Thanks for the review! I turns out I failed to do spec changes for the "Remove automatic unlocking on close/error" commit. Your let vs. set catch was correct as well. Fixed and will merge now.

**Comment by yutakahirano** [comment]
Just a confirmation: @wanderview, are you OK with this change? I am planning to implement this behavior change to Chrome, and before that I would like to make sure that everyone is satisfied.

**Comment by yutakahirano** [comment]
@wanderview: ping

**Comment by wanderview** [comment]
I'm sorry, but I haven't been able to keep up with this.  Can you summarize
the differences from the July f2f agreement?

Are we just dropping automatic unlocking on stream close?

**Comment by yutakahirano** [comment]
The f2f agreement suggests a stream should be disturbed  when read even if it is closed (See https://github.com/yutakahirano/fetch-with-streams/pull/48#issuecomment-125172550). In order to achieve that, there are two ways:

1. Auto-released reader makes the stream disturbed.
1. Remove the auto-release mechanism.

My opinion is that the latter is less evil.

> Are we just dropping automatic unlocking on stream close?

And on stream error.

**Comment by wanderview** [comment]
> Remove the auto-release mechanism.

I think this is reasonable.  Thanks for summarizing for me.
